server/src/main/java/org/elasticsearch/cluster/metadata/MetadataCreateIndexService.java

```diff
@@ -180,32 +180,21 @@ public void validateIndexName(String index, ClusterState state) {
     /**
      * Validates (if this index has a dot-prefixed name) whether it follows the rules for dot-prefixed indices.
      * @param index The name of the index in question
-     * @param state The current cluster state
      * @param isHidden Whether or not this is a hidden index
      */
-    public void validateDotIndex(String index, ClusterState state, @Nullable Boolean isHidden) {
+    public boolean validateDotIndex(String index, @Nullable Boolean isHidden) {
+        boolean isSystem = false;
         if (index.charAt(0) == '.') {
-            List<SystemIndexDescriptor> matchingDescriptors = systemIndexDescriptors.stream()
-                .filter(descriptor -> descriptor.matchesIndexPattern(index))
-                .collect(toList());
-            if (matchingDescriptors.isEmpty() && (isHidden == null || isHidden == Boolean.FALSE)) {
+            SystemIndexDescriptor matchingDescriptor = systemIndices.findMatchingDescriptor(index);
+            if (matchingDescriptor == null && (isHidden == null || isHidden == Boolean.FALSE)) {
                 deprecationLogger.deprecate("index_name_starts_with_dot",
                     "index name [{}] starts with a dot '.', in the next major version, index names " +
                     "starting with a dot are reserved for hidden indices and system indices", index);
-            } else if (matchingDescriptors.size() > 1) {
-                // This should be prevented by erroring on overlapping patterns at startup time, but is here just in case.
-                StringBuilder errorMessage = new StringBuilder()
-                    .append("index name [")
-                    .append(index)
-                    .append("] is claimed as a system index by multiple system index patterns: [")
-                    .append(matchingDescriptors.stream()
-                        .map(descriptor -> "pattern: [" + descriptor.getIndexPattern() +
-                            "], description: [" + descriptor.getDescription() + "]").collect(Collectors.joining("; ")));
-                // Throw AssertionError if assertions are enabled, or a regular exception otherwise:
-                assert false : errorMessage.toString();
-                throw new IllegalStateException(errorMessage.toString());
+            } else {
+                isSystem = true;
             }
         }
```

Comment:
This if/else, as currently written, will cause hidden indices to be marked as system indices. When a hidden index is created, `matchingDescriptor` is `null` and `isHidden` is `true`, which does not match the `if` clause, and so falls triggers the `else`, which marks the index as a system index.

Changing this to the following fixes the issue:
```java
        if (index.charAt(0) == '.') {
            SystemIndexDescriptor matchingDescriptor = systemIndices.findMatchingDescriptor(index);
            if (matchingDescriptor != null) {
                logger.trace("index [{}] is a system index because it matches index pattern [{}] with description [{}]",
                    index, matchingDescriptor.getIndexPattern(), matchingDescriptor.getDescription());
                isSystem = true;
            } else if (isHidden) {
                logger.trace("index [{}] is a hidden index", index);
            } else {
                deprecationLogger.deprecate("index_name_starts_with_dot",
                    "index name [{}] starts with a dot '.', in the next major version, index names " +
                        "starting with a dot are reserved for hidden indices and system indices", index);
            }
        }
```

Adding asserts to check the boolean returned by this method in `testValidateDotIndex` catches the issue as well.

Reply:
[Here is a commit with the fix + test adjustments](https://github.com/elastic/elasticsearch/commit/3683a3785206bc2c244559f97dffe418810ee255) if you want to cherry-pick that.